conseil-lorre/src/main/scala/tech/cryptonomic/conseil/indexer/tezos/processing/BakingAndEndorsingRightsProcessing.scala

```diff
@@ -0,0 +1,158 @@
+import scala.concurrent.Future
+
+class BakingAndEndorsingRightsProcessing(
+    db: Database,
+    configuration: config.BakingAndEndorsingRights,
+    nodeOperator: TezosNodeOperator,
+    indexedData: TezosIndexedDataOperations
+)(implicit mat: ActorMaterializer) {
+
+  private[tezos] def processBakingAndEndorsingRights(
+      fetchingResults: nodeOperator.BlockFetchingResults
+  )(
+      implicit ec: ExecutionContext
+  ): Future[Unit] = {
+    import cats.implicits._
+
+    val blockHashesWithCycleAndGovernancePeriod = fetchingResults.map { results =>
+      {
+        val data = results._1.data
+        val hash = data.hash
+        data.metadata match {
+          case GenesisMetadata => FetchRights(None, None, Some(hash))
+          case BlockHeaderMetadata(_, _, _, _, _, level) =>
+            FetchRights(Some(level.cycle), Some(level.voting_period), Some(hash))
+
+        }
+      }
+    }
+
+    (
+      nodeOperator.getBatchBakingRights(blockHashesWithCycleAndGovernancePeriod),
+      nodeOperator.getBatchEndorsingRights(blockHashesWithCycleAndGovernancePeriod)
+    ).mapN { (br, er) =>
+      val updatedEndorsingRights = updateEndorsingRights(er, fetchingResults)
+      (db.run(TezosDb.upsertBakingRights(br)), db.run(TezosDb.upsertEndorsingRights(updatedEndorsingRights)))
+    }.void
+  }
+
+  /** Updates endorsing rights with endorsed block */
+  private def updateEndorsingRights(
+      endorsingRights: Map[FetchRights, List[EndorsingRights]],
+      fetchingResults: nodeOperator.BlockFetchingResults
+  ): Map[FetchRights, List[EndorsingRights]] =
+    endorsingRights.map {
+      case (fetchRights, endorsingRightsList) =>
+        fetchRights -> endorsingRightsList.map { rights =>
+              val endorsedBlock = fetchingResults.find {
+                case (block, _) =>
+                  fetchRights.blockHash.contains(block.data.hash)
+              }.flatMap {
+                case (block, _) =>
+                  block.operationGroups.flatMap {
+                    _.contents.collect {
+                      case e: Endorsement if e.metadata.delegate.value == rights.delegate => e
+                    }.map(_.level)
+                  }.headOption
+              }
+              rights.copy(endorsedBlock = endorsedBlock)
+            }
+    }
+
+  /** Fetches future baking and endorsing rights to insert it into the DB */
+  private[tezos] def writeFutureRights()(implicit ec: ExecutionContext): Unit = {
+    val berLogger = LoggerFactory.getLogger("RightsFetcher")
+
+    import cats.implicits._
+
+    berLogger.info("Fetching future baking and endorsing rights")
+    val blockHead = nodeOperator.getBareBlockHead()
+    val brLevelFut = indexedData.fetchMaxBakingRightsLevel()
+    val erLevelFut = indexedData.fetchMaxEndorsingRightsLevel()
+
+    (blockHead, brLevelFut, erLevelFut).mapN { (head, brLevel, erLevel) =>
+      val headLevel = head.header.level
+      val rightsStartLevel = math.max(brLevel, erLevel) + 1
```

Comment:
+1 always looks suspicious to me :D

Reply:
Will check once more, I think there's nothing wrong here though

Reply:
Suspicious in a way, that sometimes it's hard do guess why you want to do +1. It can be totally valid tho.

Reply:
I still have to check this